astropy/modeling/src/projections.c.templ

```diff
@@ -0,0 +1,291 @@
+#include <Python.h> /* Python */
+
+#include <numpy/arrayobject.h> /* Numpy */
+
+#include "prj.h" /* from wcslib */
+#include "wcsmath.h"
+
+
+#if PY_MAJOR_VERSION >= 3
+#define PY3K 1
+#else
+#define PY3K 0
+#endif
+
+
+{% for name, nargs in projections.items() %}
+static PyObject*
+Py_{{ name }}x2s(PyObject* __, PyObject* args, PyObject* kwds)
+{
+  PyObject* x = NULL;
+  PyObject* y = NULL;
+  PyObject* x_array = NULL;
+  PyObject* y_array = NULL;
+  PyObject* phi_array = NULL;
+  PyObject* theta_array = NULL;
+  PyObject* stat_array = NULL;
+  Py_ssize_t n;
+  Py_ssize_t i;
+  PyObject* result = NULL;
+
+  struct prjprm prj;
+
+  int status = -1;
+
+  memset(&prj, 0, sizeof(struct prjprm));
+  prj.phi0 = UNDEFINED;
+  prj.theta0 = UNDEFINED;
+
+  if (!PyArg_ParseTuple(args,
+                        "OO"
+                        {% for i in range(nargs) %}"d" {% endfor %}
+                        ":{{ name }}x2s",
+                        &x, &y
+                        {% for i in range(nargs) %}
+                        {% if loop.first %},{% endif %}
+                        &prj.pv[{{i}} + 1]
+                        {% if not loop.last %},{% endif %}
+                        {% endfor %}
+                        )) {
+    return NULL;
+  }
+
+  /* TODO: This assumes the same shape for the input arrays.  Instead,
+     we should broadcast */
+
+  x_array = PyArray_FromObject(x, PyArray_DOUBLE, 1, NPY_MAXDIMS);
+  if (x_array == NULL) {
+    goto exit;
+  }
+
+  y_array = PyArray_FromObject(y, PyArray_DOUBLE, 1, NPY_MAXDIMS);
+  if (y_array == NULL) {
+    goto exit;
+  }
+
+  phi_array = PyArray_SimpleNew(PyArray_NDIM(x_array), PyArray_DIMS(x_array), PyArray_DOUBLE);
+  if (phi_array == NULL) {
+    goto exit;
+  }
+
+  theta_array = PyArray_SimpleNew(PyArray_NDIM(x_array), PyArray_DIMS(x_array), PyArray_DOUBLE);
+  if (theta_array == NULL) {
+    goto exit;
+  }
+
+  stat_array = PyArray_SimpleNew(PyArray_NDIM(x_array), PyArray_DIMS(x_array), PyArray_INT);
+  if (stat_array == NULL) {
+    goto exit;
+  }
+
+  if ((status = {{ name }}set(&prj))) {
+    goto exit;
+  }
+
+  n = 1;
+  for (i = 0; i < PyArray_NDIM(x_array); ++i) {
+    n *= PyArray_DIM(x_array, i);
+  }
+
+  if ((status = {{ name }}x2s(&prj, n, n,
+                       PyArray_STRIDE(x_array, PyArray_NDIM(x_array) - 1),
+                       PyArray_STRIDE(y_array, PyArray_NDIM(y_array) - 1),
+                       (double *)PyArray_DATA(x_array),
+                       (double *)PyArray_DATA(y_array),
+                       (double *)PyArray_DATA(phi_array),
+                       (double *)PyArray_DATA(theta_array),
```

Comment:
Yes.  The input arrays can be non-contiguous (that's why `PyArray_STRIDE` is passed in).  The output arrays must be contiguous, but that's less of an issue since we create/allocate them right here.